`pkgs/by-name`: Enable gradual migration checks and add `run-local.sh`

infinisil opened the pull request:
## Description of changes

After https://github.com/NixOS/nixpkgs/pull/272395 extended the `nixpkgs-check-by-name` tool to have gradual migration checking, aka [ratchet checks](https://github.com/NixOS/nixpkgs/tree/master/pkgs/test/nixpkgs-check-by-name#ratchet-checks), this PR now enables them by passing the `--base` flag to the tool.

Furthermore, this is a really good opportunity to make the CI check runnable locally, solving https://github.com/NixOS/nixpkgs/issues/266931 as well as possible, replacing https://github.com/NixOS/nixpkgs/pull/266937.
Fundamentally it's impossible to make it fully reproducible because the base branch fetching/merging can't happen in the same way, but this PR makes everything be shared that can be shared between CI and the local run.

It's not perfect, but better than nothing.

## Things done

- [x] Tested in CI: https://github.com/tweag/nixpkgs/pull/81
- [x] Tested locally:
  ```
  $ ./pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh by-name-reproducible https://github.com/tweag/nixpkgs.git
  Using HEAD commit cf06cb1cd4cf0f0e92fa3bd3ce894c1255fae8a1
  Creating Git worktree for the HEAD commit in /tmp/tmp.osLpsrjqvl/merged.. Done
  Fetching base branch by-name-reproducible to compare against.. cc0fe3b4b9f268e53fd0c1ce706f91a8d6af1373
  Creating Git worktree for the base branch in /tmp/tmp.osLpsrjqvl/base.. Done
  Merging base branch into the HEAD commit in /tmp/tmp.osLpsrjqvl/merged.. cf06cb1cd4cf0f0e92fa3bd3ce894c1255fae8a1
  Determining the channel to use for PR base branch by-name-reproducible.. nixos-unstable
  Fetching latest version of channel nixos-unstable.. /nix/store/k87dn8r5rr9nsa297my0h3z02hx62mnq-source
  Git revision of channel nixos-unstable is 91a00709aebb3602f172a0bf47ba1ef013e34835
  Fetching the prebuilt version of nixpkgs-check-by-name.. /nix/store/381kv4kh7sx2a43w8vcbqcy67msvyfzq-nixpkgs-check-by-name
  Running nixpkgs-check-by-name..
  pkgs.vcpkg-tool: This attribute is manually defined (most likely in pkgs/top-level/all-packages.nix), which is only allowed if the definition is of the form `pkgs.callPackage pkgs/by-name/vc/vcpkg-tool/package.nix { ... }` with a non-empty second argument.
  Validation failed, see above errors
  Cleaning up.. Done
  ```
- [x] Updated contributor docs and code comments
- [x] Clean commit history

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
workflows/check-by-name: Move tool fetching into script

This part of the CI can also be reproduced locally

infinisil committed:
workflows/check-by-name: Why the mergeability check needs to be inline

infinisil committed:
tests.nixpkgs-check-by-name: Create script to run locally

Due to the check soon depending on the base branch (see `--base`),
the CI check can't reasonably share all code with a local check.
We can still make a script to run it locally, just not sharing all code.

infinisil committed:
workflows/check-by-name: Slim down and prepare for --base

Now that we have a script to run the check locally,
there's no real need to output the information to reproduce anymore,
which allows cleaning up the CI workflow.

Furthermore, this prepares the CI workflow to be passed `--base`, as
introduced recently.

infinisil committed:
check-by-name: Pass --base in CI and local running script

This enables the ratchet checks for pkgs/by-name, allowing gradual
migrations!

infinisil committed:
tests.nixpkgs-check-by-name: Add documentation for scripts

roberth commented on pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh:
```diff
@@ -0,0 +1,66 @@
+echo >&2 "Done"
+
+echo >&2 -n "Merging base branch into the HEAD commit in $tmp/merged.. "
+git -C "$tmp/merged" merge -q --no-edit "$baseSha" 2>/dev/null
```
Please don't throw away the error message.

If you want fancy logging, you must capture stderr, but it's a waste of effort.

infinisil replied:
Oh yeah I disabled it originally because the messages can get really long if files need to be merged, but this isn't a problem in practice. I removed the pipe to `/dev/null` now.

RaitoBezarius commented on pkgs/by-name/README.md:
````diff
@@ -112,12 +112,11 @@ There's some limitations as to which packages can be defined using this structur
-nix-build -A tests.nixpkgs-check-by-name
-result/bin/nixpkgs-check-by-name .
 ```
+$ ./pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh master
````
Why not put it inside `maintainers/scripts` ?

infinisil replied:
`run-local.sh` is indeed a script only for maintainers, but there's also `fetch-tool.sh` next to it, which is used by both `run-local.sh` and CI, so that wouldn't strictly be a maintainer script.

I guess `run-local.sh` should be in `maintainers/scripts`, while `fetch-tool.sh` could stay in `pkgs/test/nixpkgs-check-by-name/scripts`, though then they're a bit far apart, and the `pkgs/by-name` implementation is already split into like 5 different places.

roberth replied:
I'm all for keeping things together.
If we nonetheless insist on having `maintainers/scripts`, it could be a minimal wrapper that invokes the real thing.

RaitoBezarius replied:
Sounds like a symlink cost nothing here?

infinisil replied:
Yeah that's an idea. I'll do the symlink and make its path be printed in CI.

infinisil replied:
Now introduced `maintainers/scripts/check-by-name.sh` as a symlink in https://github.com/NixOS/nixpkgs/pull/274591/commits/c106c49923fa7cf656ea464b59f74ef623966e82

RaitoBezarius replied:
Good to me

l0b0 commented on pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh:
```diff
@@ -0,0 +1,66 @@
+
+cleanup_commands=()
+cleanup() {
+    echo -n >&2 "Cleaning up.. "
```
Putting the redirect at the end of the command is more idiomatic. It doesn't belong as an argument.

infinisil replied:
I kind of don't like that idiom, because it leads to `>&2` being all over the place, whereas putting it in front makes them all be fairly aligned. I guess I could declare a function like `debug() { echo >&2 "$@" }` to not have it all over the place.

l0b0 replied:
A `debug` function would be preferable. But I think idiomaticity is more important than aligning things vertically. Having `COMMAND OPTIONS PARAMETERS REDIRECT` or `REDIRECT COMMAND OPTIONS PARAMETERS` is cleaner than `COMMAND OPTIONS REDIRECT OPTIONS PARAMETERS`.

infinisil replied:
Now introduced a `trace` (sounded better than `debug`, since it's not strictly for debugging) function for this purpose in https://github.com/NixOS/nixpkgs/pull/274591/commits/ca7201d88358394118120162f19bc75c7fec5a13

l0b0 commented on pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh:
```diff
@@ -0,0 +1,66 @@
+    echo -n >&2 "Cleaning up.. "
+    # Run all cleanup commands in inverse order
+    for (( i=${#cleanup_commands[@]}-1; i>=0; i-- )); do
+        eval "${cleanup_commands[i]}"
```
This poses a lot of danger. Are there actually any commands which need to be `eval`ed here? The ones below all seem to refer to variables which are already in scope, so we could just do

```
cleanup_commands+=("git worktree remove --force '$tmp/merged'")
```

and then

```
"${cleanup_commands[i]}"
```

in this line.

infinisil replied:
That doesn't work because it would try to execute `git worktree remove --force '$tmp/merged'` as a file (just tried it out).

l0b0 replied:
My bad, nested quotes always get me. Would it be possible to write the cleanup commands to a temporary file? That should be safer than `eval`.

infinisil replied:
Are you thinking of something like this?
```bash
touch foo
echo "rm foo" >> "$tmp/cleanup_commands"

source "$tmp/cleanup_commands"
```

That sounds just like the `eval` approach, just with one more level of indirection.

---

Is there really a safety risk with `eval` here? It seems fine to me.

l0b0 replied:
`eval` is the biggest footgun in Bash. If anyone changes any of the cleanup commands to anything even slightly more complex in the future, the potential for disaster is there.

infinisil replied:
We discussed this [on Matrix](https://matrix.to/#/!kjdutkOsheZdjqYmqp:nixos.org/$XmMLYPzkkX2ddFYTxGBTrZBDm7Yz92ggmK_IXGOLRnY?via=nixos.org&via=matrix.org&via=nixos.dev). I updated the script accordingly now: b62b605c23b6ff7771ac8bf9a2c78ee2ddd69b14

infinisil committed:
pkgs/test/nixpkgs-check-by-name/scripts: Various improvements

- trace function, avoids littering `echo >&2` all throughout
- Avoid `eval`, remove unneeded shellcheck

Co-Authored-By: Victor Engmark <victor@engmark.name>

infinisil committed:
maintainers/scripts/check-by-name.sh: Introduce symlink alias

l0b0 commented on pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh:
```diff
@@ -0,0 +1,68 @@
+trap cleanup exit
+
+tmp=$(mktemp -d)
+cleanup_commands+=('rmdir "$tmp"')
```
The cleanup commands should generally be set up *before* the command which creates the resource, so that if the resource is created but the command returns a non-zero exit code for some other reason the cleanup still happens.

infinisil replied:
Ah good point! Done in 206361ab3889aeaee4428d8cd2e175cbdc5fbdc7

l0b0 commented on pkgs/test/nixpkgs-check-by-name/scripts/run-local.sh:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env bash
+# shellcheck disable=SC2016
```
I think you don't need this anymore

infinisil replied:
Now in https://github.com/NixOS/nixpkgs/pull/274591/commits/e130ee33a11ae90585e9f9c09b189c3ccfe70632 :)

infinisil commented:
Thanks for the detailed review @l0b0! I now squashed all the script improvements into https://github.com/NixOS/nixpkgs/pull/274591/commits/e130ee33a11ae90585e9f9c09b189c3ccfe70632, adding you as a co-author :)

infinisil commented:
I'll merge this, because:
- I consider https://github.com/NixOS/nixpkgs/pull/274591#discussion_r1434415794 resolved, and it's a minor comment anyways, and we can always improve this as we go
- I tested the workflow (https://github.com/tweag/nixpkgs/pull/81) and script again, all still works
- We don't need to wait for ofborg because this doesn't touch any packages

This allows me to proceed with finalising the next step for RFC 140:

- https://github.com/NixOS/nixpkgs/pull/275539
